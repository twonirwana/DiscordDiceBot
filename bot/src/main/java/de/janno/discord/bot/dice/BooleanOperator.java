package de.janno.discord.bot.dice;

import java.util.Objects;
import java.util.function.BiFunction;

public enum BooleanOperator {
    GREATER(">", ">", (a, b) -> a > b),
    GREATER_EQUAL(">=", "≥", (a, b) -> a >= b),
    LESSER("<", "<", (a, b) -> a < b),
    LESSER_EQUAL("<=", "≤", (a, b) -> a <= b),
    EQUAL("=", "=", Objects::equals),
    NOT_EQUAL("<>", "≠", (a, b) -> !Objects.equals(a, b));

    final String expression;
    final String pretty;
    final BiFunction<Integer, Integer, Boolean> function;
    BooleanOperator(String expression, String pretty, BiFunction<Integer, Integer, Boolean> function) {
        this.expression = expression;
        this.pretty = pretty;
        this.function = function;
    }
}
