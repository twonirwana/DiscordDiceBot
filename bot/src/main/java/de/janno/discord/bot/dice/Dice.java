package de.janno.discord.bot.dice;

import dev.diceroll.parser.ResultTree;

public interface Dice {
    ResultTree detailedRoll(String input);
}
