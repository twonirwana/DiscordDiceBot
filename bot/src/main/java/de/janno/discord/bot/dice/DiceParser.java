package de.janno.discord.bot.dice;

import dev.diceroll.parser.ResultTree;

public class DiceParser implements Dice {

    @Override
    public ResultTree detailedRoll(String input) {
        return dev.diceroll.parser.Dice.detailedRoll(input);
    }
}
