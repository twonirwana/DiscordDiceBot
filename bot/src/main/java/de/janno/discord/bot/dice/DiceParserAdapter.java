package de.janno.discord.bot.dice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableList;
import de.janno.discord.bot.command.AnswerFormatType;
import de.janno.discord.bot.command.RollAnswer;
import de.janno.discord.connector.api.BottomCustomIdUtils;
import dev.diceroll.parser.ResultTree;
import lombok.NonNull;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Slf4j
public class DiceParserAdapter {

    private static final Pattern BOOLEAN_EXPRESSION_PATTERN = Pattern.compile("(^.+?)((?:<=|>=|<>|<|>|=)\\d+\\?.+)+:(.+)$");
    private static final Pattern MULTI_ROLL_EXPRESSION_PATTERN = Pattern.compile("^(\\d+?)x\\[(.*)?]$");
    private static final Pattern VALUE_COMPERE_PATTER = Pattern.compile("(<=|>=|<>|<|>|=)(\\d+)\\?(.+)");
    private static final String MULTI_DIFF_EXPRESSION_DELIMITER = "&";
    private final Dice dice;

    public DiceParserAdapter() {
        this(new DiceParser());
    }

    @VisibleForTesting
    public DiceParserAdapter(Dice dice) {
        this.dice = dice;
    }

    @VisibleForTesting
    static boolean isMultipleRoll(String input) {
        if (input.contains("x[") && input.contains(MULTI_DIFF_EXPRESSION_DELIMITER)) {
            return false;
        }
        return isMultipleIdenticalRolls(input)
                || isMultipleDifferentRolls(input);
    }

    @VisibleForTesting
    static boolean isMultipleIdenticalRolls(String input) {
        return MULTI_ROLL_EXPRESSION_PATTERN.matcher(input).matches();
    }

    @VisibleForTesting
    static boolean isMultipleDifferentRolls(String input) {
        return input.contains(MULTI_DIFF_EXPRESSION_DELIMITER);
    }

    @VisibleForTesting
    static boolean isBooleanExpression(String input) {
        return BOOLEAN_EXPRESSION_PATTERN.matcher(input).matches();
    }

    @VisibleForTesting
    static int getNumberOfMultipleRolls(String input) {
        Matcher matcher = MULTI_ROLL_EXPRESSION_PATTERN.matcher(input);
        if (matcher.find()) {
            int numberOfRolls = Integer.parseInt(matcher.group(1));
            return Math.min(numberOfRolls, 25); //limited to 25 because that is the max number of embed discord fields
        }
        throw new IllegalArgumentException(String.format("Number of multiplier in '%s' not found", input));

    }

    @VisibleForTesting
    static String getInnerDiceExpressionFromMultiRoll(String input) {
        Matcher matcher = MULTI_ROLL_EXPRESSION_PATTERN.matcher(input);
        if (matcher.find()) {
            return matcher.group(2);
        }
        throw new IllegalArgumentException(String.format("Inner expression in '%s' not found", input));
    }

    private static List<Integer> getBaseResults(ResultTree resultTree) {
        if (!resultTree.getResults().isEmpty()) {
            return resultTree.getResults().stream()
                    .flatMap(rt -> getBaseResults(rt).stream())
                    .collect(Collectors.toList());
        }
        return ImmutableList.of(resultTree.getValue());
    }

    public static @NonNull String getExpressionFromExpressionWithOptionalLabel(String expressionWithOptionalLabel, String labelDelimiter) {
        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            int firstDelimiter = expressionWithOptionalLabel.indexOf(labelDelimiter);
            return expressionWithOptionalLabel.substring(0, firstDelimiter);
        }
        return expressionWithOptionalLabel;
    }

    public static Optional<String> getLabelFromExpressionWithOptionalLabel(String expressionWithOptionalLabel, String labelDelimiter) {
        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            int firstDelimiter = expressionWithOptionalLabel.indexOf(labelDelimiter);
            return Optional.of(expressionWithOptionalLabel.substring(firstDelimiter + labelDelimiter.length()));
        }
        return Optional.empty();
    }

    public Optional<String> validateDiceExpression(String expression) {
        if (!validExpression(expression)) {
            return Optional.of(String.format("The following dice expression is invalid: '%s'", expression));
        }
        return Optional.empty();
    }

    public Optional<String> validateDiceExpressionWitOptionalLabel(@NonNull String expressionWithOptionalLabel, String labelDelimiter) {
        String label;
        String diceExpression;

        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            String[] split = expressionWithOptionalLabel.split(labelDelimiter);
            if (split.length != 2) {
                return Optional.of(String.format("The button definition '%s' should have the diceExpression@Label", expressionWithOptionalLabel));
            }
            label = split[1].trim();
            diceExpression = split[0].trim();
        } else {
            label = expressionWithOptionalLabel;
            diceExpression = expressionWithOptionalLabel;
        }
        if (label.length() > 80) {
            return Optional.of(String.format("Label for '%s' is to long, max number of characters is 80", expressionWithOptionalLabel));
        }
        if (label.isBlank()) {
            return Optional.of(String.format("Label for '%s' requires a visible name", expressionWithOptionalLabel));
        }
        if (diceExpression.isBlank()) {
            return Optional.of(String.format("Dice expression for '%s' is empty", expressionWithOptionalLabel));
        }
        return validateDiceExpression(diceExpression);
    }

    public Optional<String> validateListOfExpressions(List<String> optionValues, String labelDelimiter) {
        if (optionValues.isEmpty()) {
            return Optional.of("You must configure at least one dice expression");
        }
        for (String startOptionString : optionValues) {
            if (startOptionString.contains(BottomCustomIdUtils.CUSTOM_ID_DELIMITER)) {
                return Optional.of(String.format("The button definition '%s' is not allowed to contain '%s'", startOptionString, BottomCustomIdUtils.CUSTOM_ID_DELIMITER));
            }
            Optional<String> diceParserValidation = validateDiceExpressionWitOptionalLabel(startOptionString, labelDelimiter);
            if (diceParserValidation.isPresent()) {
                return diceParserValidation;
            }
        }

        return Optional.empty();
    }

    private String removeLeadingPlus(String diceExpression) {
        if (diceExpression.startsWith("+")) {
            return diceExpression.substring(1);
        }
        return diceExpression;
    }

    private List<String> splitMultipleDifferentExpressions(String input) {
        return Arrays.stream(input.split(MULTI_DIFF_EXPRESSION_DELIMITER))
                .map(String::trim)
                .filter(s -> !Strings.isNullOrEmpty(s))
                .collect(Collectors.toList());
    }

    public RollAnswer answerRollWithOptionalLabelInExpression(String expressionAndOptionalLabel, String labelDelimiter, AnswerFormatType answerFormatType) {
        String label = DiceParserAdapter.getLabelFromExpressionWithOptionalLabel(expressionAndOptionalLabel, labelDelimiter).orElse(null);
        String diceExpression = DiceParserAdapter.getExpressionFromExpressionWithOptionalLabel(expressionAndOptionalLabel, labelDelimiter);
        return answerRollWithGivenLabel(diceExpression, label, answerFormatType);
    }

    public RollAnswer answerRollWithGivenLabel(String input, @Nullable String label, AnswerFormatType answerFormatType) {
        try {
            if (isMultipleRoll(input)) {
                List<ExpressionLabelResultAndDetails> expressionLabelResultAndDetails;
                if (isMultipleIdenticalRolls(input)) {
                    int numberOfRolls = getNumberOfMultipleRolls(input);
                    String innerExpression = getInnerDiceExpressionFromMultiRoll(input);
                    expressionLabelResultAndDetails = IntStream.range(0, numberOfRolls)
                            .mapToObj(i -> singleRoll(innerExpression, null))
                            .collect(Collectors.toList());
                } else if (isMultipleDifferentRolls(input)) {
                    expressionLabelResultAndDetails = splitMultipleDifferentExpressions(input).stream()
                            .map(s -> singleRoll(s, null))
                            .collect(Collectors.toList());
                } else {
                    throw new IllegalStateException(String.format("Can't find correct multi roll version for: %s", input));
                }

                List<RollAnswer.RollResults> multiRollResults = expressionLabelResultAndDetails.stream()
                        .map(r -> new RollAnswer.RollResults(r.getExpression(), r.getResult(), r.getDetails()))
                        .collect(ImmutableList.toImmutableList());
                return RollAnswer.builder()
                        .answerFormatType(answerFormatType)
                        .expression(input)
                        .expressionLabel(Strings.isNullOrEmpty(label) ? "Multiple Results" : label)
                        .multiRollResults(multiRollResults)
                        .build();
            } else {
                ExpressionLabelResultAndDetails expressionLabelResultAndDetails = singleRoll(input, label);
                return RollAnswer.builder()
                        .answerFormatType(answerFormatType)
                        .expression(expressionLabelResultAndDetails.getExpression())
                        .expressionLabel(expressionLabelResultAndDetails.getLabel())
                        .result(expressionLabelResultAndDetails.getResult())
                        .rollDetails(expressionLabelResultAndDetails.getDetails())
                        .build();
            }
        } catch (Throwable t) {
            return RollAnswer.builder()
                    .answerFormatType(answerFormatType)
                    .expression(input)
                    .errorMessage(t.getMessage())
                    .build();
        }
    }

    private ExpressionLabelResultAndDetails singleRoll(String input, String label) {
        if (isBooleanExpression(input)) {
            BooleanExpression booleanExpression = getBooleanExpression(input);
            RollWithDetails rollWithDetails = rollWithDiceParser(booleanExpression.getExpression());
            if (rollWithDetails.getResult() == null) { //there was an error
                return new ExpressionLabelResultAndDetails(input, label, rollWithDetails.getRoll(), rollWithDetails.getDetails());
            }
            String result = booleanExpression.getResult(rollWithDetails.getResult());
            String details = String.format("%s = %s", rollWithDetails.getDetails(), booleanExpression.getDetail(rollWithDetails.getResult()));
            return new ExpressionLabelResultAndDetails(booleanExpression.getExpression(), label, result, details);
        } else {
            RollWithDetails rollWithDetails = rollWithDiceParser(input);
            return new ExpressionLabelResultAndDetails(input, label, String.valueOf(rollWithDetails.getResult()), rollWithDetails.getDetails());
        }
    }

    @VisibleForTesting
    BooleanExpression getBooleanExpression(String input) {
        Matcher matcher = BOOLEAN_EXPRESSION_PATTERN.matcher(input);
        if (!matcher.find()) {
            throw new IllegalArgumentException(String.format("'%s' doesn't match the required patter '%s'", input, BOOLEAN_EXPRESSION_PATTERN));
        }
        String expression = matcher.group(1);
        String compares = matcher.group(2);
        String defaultAnswer = matcher.group(3);
        List<ValueCompereResult> valueCompereResults = Arrays.stream(compares.split("(?<=[^<>=])(?=<=|>=|<>|<|>|=)"))
                .map(this::parseValueCompereResult)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
        return new BooleanExpression(expression, valueCompereResults, defaultAnswer);
    }

    private ValueCompereResult parseValueCompereResult(String in) {
        Matcher matcher = VALUE_COMPERE_PATTER.matcher(in);
        if (!matcher.find()) {
            return null;
        }
        String booleanOperatorExpression = matcher.group(1);
        BooleanOperator operator = Arrays.stream(BooleanOperator.values())
                .filter(o -> o.expression.equals(booleanOperatorExpression))
                .findFirst()
                .orElseThrow(() -> new IllegalStateException(String.format("'%s' is not a valid boolean operator ", booleanOperatorExpression)));
        int compereValue = Integer.parseInt(matcher.group(2));
        String result = matcher.group(3);
        return new ValueCompereResult(operator, compereValue, result);
    }

    private RollWithDetails rollWithDiceParser(String input) {
        input = removeLeadingPlus(input);
        ResultTree resultTree = dice.detailedRoll(input);
        String title = String.format("%s = %d", input, resultTree.getValue());
        String details = String.format("[%s]", getBaseResults(resultTree).stream().map(String::valueOf).collect(Collectors.joining(", ")));
        return new RollWithDetails(title, details, resultTree.getValue());
    }

    public boolean validExpression(String input) {
        try {
            input = removeLeadingPlus(input);
            if (isMultipleRoll(input)) {
                if (isMultipleIdenticalRolls(input)) {
                    singleRoll(getInnerDiceExpressionFromMultiRoll(input), null);
                } else if (isMultipleDifferentRolls(input)) {
                    splitMultipleDifferentExpressions(input).forEach(e -> singleRoll(e, null));
                }
            } else if (isBooleanExpression(input)) {
                dice.detailedRoll(getBooleanExpression(input).getExpression());
            } else {
                dice.detailedRoll(input);
            }
            return true;
        } catch (Throwable t) {
            return false;
        }
    }

    @Value
    private static class ExpressionLabelResultAndDetails {
        @NonNull
        String expression;
        @Nullable
        String label;
        @NonNull
        String result;
        @NonNull
        String details;
    }

    @Value
    private static class RollWithDetails {
        @NonNull
        String roll;
        @NonNull
        String details;
        Integer result;
    }

    @Value
    @VisibleForTesting
    static class BooleanExpression {
        String expression;
        List<ValueCompereResult> valueCompereResults;
        String defaultResult;

        public String getResult(int value) {
            return valueCompereResults.stream()
                    .filter(vcr -> vcr.matches(value))
                    .findFirst()
                    .map(ValueCompereResult::getResultValue)
                    .orElse(defaultResult);
        }

        public String getDetail(int value) {
            String resultDetail = valueCompereResults.stream()
                    .filter(vcr -> vcr.matches(value))
                    .findFirst()
                    .map(ValueCompereResult::toString)
                    .orElse(String.format(" ⟹ %s", defaultResult));
            return String.format("%d%s", value, resultDetail);
        }
    }

    @Value
    @VisibleForTesting
    static class ValueCompereResult {

        BooleanOperator operator;
        int compareValue;
        String resultValue;

        private boolean matches(int value) {
            return operator.function.apply(value, compareValue);
        }

        public String toString() {
            return String.format("%s%d ⟹ %s", operator.pretty, compareValue, resultValue);
        }
    }

}
