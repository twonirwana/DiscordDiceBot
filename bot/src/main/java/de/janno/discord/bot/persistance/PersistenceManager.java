package de.janno.discord.bot.persistance;

import lombok.NonNull;

import java.util.Optional;
import java.util.Set;
import java.util.UUID;

public interface PersistenceManager {

    @NonNull Optional<MessageConfigDTO> getMessageConfig(@NonNull UUID configUUID);

    @NonNull Optional<MessageConfigDTO> getConfigFromMessage(long channelId, long messageId);

    void saveMessageConfig(@NonNull MessageConfigDTO messageConfigDTO);

    @NonNull Optional<MessageDataDTO> getMessageData(long channelId, long messageId);

    @NonNull Set<Long> getAllMessageIdsForConfig(@NonNull UUID configUUID);

    void deleteStateForMessage(long channelId, long messageId);

    @NonNull Set<Long> deleteMessageDataForChannel(long channelId);

    void saveMessageData(@NonNull MessageDataDTO messageState);

    Set<Long> getAllGuildIds();

    @NonNull Optional<ChannelConfigDTO> getChannelConfig(long channelId, String configClassId);

    void saveChannelConfig(@NonNull ChannelConfigDTO channelConfigDTO);

    void deleteChannelConfig(long channelId, String configClassId);
}
