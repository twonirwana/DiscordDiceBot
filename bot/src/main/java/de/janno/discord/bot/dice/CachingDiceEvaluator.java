package de.janno.discord.bot.dice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.ExpressionException;
import de.janno.evaluator.dice.Roller;
import de.janno.evaluator.dice.random.NumberSupplier;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.Tags;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.ExecutionException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

@Slf4j
public class CachingDiceEvaluator {

    private final LoadingCache<String, RollerOrError> diceRollerCache;
    private final DiceEvaluator diceEvaluator;

    public CachingDiceEvaluator(NumberSupplier numberSupplier, int maxNumberOfDice, int cacheSize) {
        diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
        diceRollerCache = CacheBuilder.newBuilder()
                .maximumSize(cacheSize)
                .recordStats()
                .build(new CacheLoader<>() {
                    @Override
                    public @NonNull RollerOrError load(@NonNull String expression) {
                        try {
                            log.debug("create roller for: {}", expression);
                            Roller roller = diceEvaluator.buildRollSupplier(expression);
                            roller.roll();
                            return new RollerOrError(expression, roller, true, null);
                        } catch (ExpressionException | ArithmeticException e) {
                            return new RollerOrError(expression, null, false, e.getMessage());
                        }
                    }
                });
        Gauge.builder("diceEvaluator.cache", diceRollerCache::size).tags(Tags.of("stats", "size")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().requestCount()).tags(Tags.of("stats", "requests")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().hitCount()).tags(Tags.of("stats", "hit")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().missCount()).tags(Tags.of("stats", "miss")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().totalLoadTime()).tags(Tags.of("stats", "totalLoadTime")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().averageLoadPenalty()).tags(Tags.of("stats", "averageLoadTime")).register(globalRegistry);
    }

    public @NonNull RollerOrError get(@NonNull String expression) {
        try {
            return diceRollerCache.get(expression);
        } catch (ExecutionException e) {
            throw new IllegalStateException(e);
        }
    }

    @VisibleForTesting
    public long getCacheSize() {
        return diceRollerCache.size();
    }
}
