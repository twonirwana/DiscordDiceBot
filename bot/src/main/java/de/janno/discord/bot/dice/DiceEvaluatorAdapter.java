package de.janno.discord.bot.dice;

import com.google.common.collect.ImmutableList;
import de.janno.discord.connector.api.message.EmbedDefinition;
import de.janno.evaluator.ExpressionException;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.random.NumberSupplier;
import de.janno.evaluator.dice.Roll;
import de.janno.evaluator.dice.RollElement;
import lombok.NonNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Optional;

public class DiceEvaluatorAdapter {

    private final DiceEvaluator diceEvaluator;

    public DiceEvaluatorAdapter(NumberSupplier numberSupplier, int maxNumberOfDice) {
        this.diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
    }

    private static @NonNull String getExpressionFromExpressionWithOptionalLabel(String expressionWithOptionalLabel, String labelDelimiter) {
        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            int firstDelimiter = expressionWithOptionalLabel.indexOf(labelDelimiter);
            return expressionWithOptionalLabel.substring(0, firstDelimiter);
        }
        return expressionWithOptionalLabel;
    }

    private static Optional<String> getLabelFromExpressionWithOptionalLabel(String expressionWithOptionalLabel, String labelDelimiter) {
        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            int firstDelimiter = expressionWithOptionalLabel.indexOf(labelDelimiter);
            String label = expressionWithOptionalLabel.substring(firstDelimiter + labelDelimiter.length());
            if (label.length() > 0) {
                return Optional.of(label);
            }
        }
        return Optional.empty();
    }

    private static String getTitleResult(Roll roll, boolean sumUp) {
        if (sumUp && allElementsAreIntegers(roll) && allElementsHaveNoColor(roll)) {
            return String.valueOf(roll.getElements().stream().flatMap(r -> r.asInteger().stream()).mapToInt(i -> i).sum());
        }
        return roll.getResultString();
    }

    private static boolean allElementsAreIntegers(Roll roll) {
        return roll.getElements().stream().allMatch(r -> r.asInteger().isPresent());
    }

    private static boolean allElementsHaveNoColor(Roll roll) {
        return roll.getElements().stream().allMatch(r -> RollElement.NO_COLOR.equals(r.getColor()));
    }

    private static String getDetailResult(@NonNull Roll result, @Nullable String expression) {
        return Optional.ofNullable(expression).map(e -> "%s throws %s".formatted(e, result.getRandomElementsString())).orElse(result.getRandomElementsString());
    }

    public static String getHelp() {
        return "```\n" + DiceEvaluator.getHelpText() + "\n```\nSee here: https://github.com/twonirwana/DiceEvaluator";
    }

    public Optional<String> validateDiceExpression(String expression, String helpCommand) {
        try {
            diceEvaluator.evaluate(expression);
            return Optional.empty();
        } catch (ExpressionException | ArithmeticException e) {
            return Optional.of(String.format("The following expression is invalid: '%s'. The error is: %s. Use %s to get more information on how to use the command.", expression, e.getMessage(), helpCommand));
        }
    }

    public EmbedDefinition answerRollWithOptionalLabelInExpression(String expression, String labelDelimiter, boolean sumUp) {
        String diceExpression = getExpressionFromExpressionWithOptionalLabel(expression, labelDelimiter);
        String label = getLabelFromExpressionWithOptionalLabel(expression, labelDelimiter).orElse(null);
        return answerRollWithGivenLabel(diceExpression, label, sumUp);
    }

    public EmbedDefinition answerRollWithGivenLabel(String diceExpression, @Nullable String label, boolean sumUp) {
        EmbedDefinition answer;
        Optional<String> optionalLabel = Optional.ofNullable(label);
        try {
            List<Roll> rolls = diceEvaluator.evaluate(diceExpression);
            if (rolls.size() == 1) {
                String title = optionalLabel.orElse(diceExpression);
                answer = EmbedDefinition.builder()
                        .title("%s ⇒ %s".formatted(title, getTitleResult(rolls.get(0), sumUp)))
                        .description(getDetailResult(rolls.get(0), optionalLabel.map(l -> diceExpression).orElse(null)))
                        .build();
            } else {
                List<EmbedDefinition.Field> fields = rolls.stream()
                        .limit(25) //max number of embedFields
                        .map(r -> new EmbedDefinition.Field("%s ⇒ %s".formatted(r.getExpression(), getTitleResult(r, sumUp)), getDetailResult(r, null), false))
                        .collect(ImmutableList.toImmutableList());
                answer = EmbedDefinition.builder()
                        .title(optionalLabel.orElse(diceExpression))
                        .fields(fields)
                        .build();
            }
        } catch (ExpressionException e) {
            answer = EmbedDefinition.builder()
                    .title("Error in: " + diceExpression)
                    .description(e.getMessage())
                    .build();
        }
        return answer;
    }


    public boolean validExpression(String expression) {
        try {
            diceEvaluator.evaluate(expression);
            return true;
        } catch (ExpressionException | ArithmeticException e) {
            return false;
        }
    }
}
