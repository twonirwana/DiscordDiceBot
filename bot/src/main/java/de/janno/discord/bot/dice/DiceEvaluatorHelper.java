package de.janno.discord.bot.dice;

import com.google.common.collect.ImmutableList;
import de.janno.discord.connector.api.message.EmbedDefinition;
import de.janno.evaluator.ExpressionException;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.NumberSupplier;
import de.janno.evaluator.dice.Result;
import de.janno.evaluator.dice.ResultElement;
import lombok.NonNull;

import java.util.List;
import java.util.Optional;

public class DiceEvaluatorHelper {

    private final DiceEvaluator diceEvaluator;

    public DiceEvaluatorHelper(NumberSupplier numberSupplier, int maxNumberOfDice) {
        this.diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
    }

    private static @NonNull String getExpressionFromExpressionWithOptionalLabel(String expressionWithOptionalLabel, String labelDelimiter) {
        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            int firstDelimiter = expressionWithOptionalLabel.indexOf(labelDelimiter);
            return expressionWithOptionalLabel.substring(0, firstDelimiter);
        }
        return expressionWithOptionalLabel;
    }

    private static Optional<String> getLabelFromExpressionWithOptionalLabel(String expressionWithOptionalLabel, String labelDelimiter) {
        if (expressionWithOptionalLabel.contains(labelDelimiter)) {
            int firstDelimiter = expressionWithOptionalLabel.indexOf(labelDelimiter);
            String label = expressionWithOptionalLabel.substring(firstDelimiter + labelDelimiter.length());
            if (label.length() > 0) {
                return Optional.of(label);
            }
        }
        return Optional.empty();
    }

    private static String getTitleResult(Result result) {
        if (result.getElements().stream().allMatch(r -> r.asInteger().isPresent()) &&
                result.getElements().stream().allMatch(r -> ResultElement.NO_COLOR.equals(r.getColor()))) {
            return String.valueOf(result.getElements().stream().flatMap(r -> r.asInteger().stream()).mapToInt(i -> i).sum());
        } else {
            return result.getResultString();
        }
    }

    private static String getDetailResult(Result result) {
        return result.getRandomElementsString();
    }

    public EmbedDefinition answerRoll(String expression, String labelDelimiter) {
        EmbedDefinition answer;
        String diceExpression = getExpressionFromExpressionWithOptionalLabel(expression, labelDelimiter);
        Optional<String> label = getLabelFromExpressionWithOptionalLabel(expression, labelDelimiter);
        try {
            List<Result> results = diceEvaluator.evaluate(diceExpression);
            if (results.size() == 1) {
                String title = label.map(l -> String.format("%s: %s", l, diceExpression)).orElse(diceExpression);
                answer = EmbedDefinition.builder()
                        .title("%s = %s".formatted(title, getTitleResult(results.get(0))))
                        .description(getDetailResult(results.get(0)))
                        .build();
            } else {
                List<EmbedDefinition.Field> fields = results.stream()
                        .limit(25) //max number of embedFields
                        .map(r -> new EmbedDefinition.Field(r.getExpression() + " = " + getTitleResult(r), getDetailResult(r), false))
                        .collect(ImmutableList.toImmutableList());
                answer = EmbedDefinition.builder()
                        .title(label.orElse(diceExpression))
                        .fields(fields)
                        .build();
            }
        } catch (ExpressionException e) {
            answer = EmbedDefinition.builder()
                    .title("Error in: " + diceExpression)
                    .description(e.getMessage())
                    .build();
        }
        return answer;
    }

    public static String getHelp() {
        return "See here: https://github.com/twonirwana/DiceEvaluator";
    }
}
