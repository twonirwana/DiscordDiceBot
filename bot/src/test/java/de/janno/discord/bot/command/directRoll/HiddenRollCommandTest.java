package de.janno.discord.bot.command.directRoll;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.persistance.PersistenceManager;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import static org.mockito.Mockito.mock;

@ExtendWith(SnapshotExtension.class)
public class HiddenRollCommandTest {

    HiddenDirectRollCommand underTest;
    private Expect expect;


    @BeforeEach
    void setup() {
        underTest = new HiddenDirectRollCommand(mock(PersistenceManager.class), new CachingDiceEvaluator((minExcl, maxIncl) -> 1, 1000, 0));
    }

    @Test
    public void getCommandDefinition() {
        expect.toMatchSnapshot(underTest.getCommandDefinition());
    }

    @Test
    public void getId() {
        expect.toMatchSnapshot(underTest.getCommandId());
    }
}
