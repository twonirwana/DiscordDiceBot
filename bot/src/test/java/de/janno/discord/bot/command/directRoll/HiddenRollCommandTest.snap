de.janno.discord.bot.command.directRoll.HiddenRollCommandTest.getCommandDefinition=[
  {
    "name": "h",
    "description": "hidden direct roll of dice expression, e.g. `2d6`. Configuration with `/channel_config`",
    "nameLocales": [ ],
    "descriptionLocales": [
      {
        "locale": "de",
        "description": "Versteckter direkter Wurf eines W체rfelausdruckes, z.B. `2d6`. Konfiguration mit `/channel_config`"
      },
      {
        "locale": "pt_BR",
        "description": "f처rmula de rolagem de dados direta escondida, e.g. `2d6`. Configure com `/channel_config`"
      }
    ],
    "options": [
      {
        "type": "STRING",
        "name": "expression",
        "nameLocales": [
          {
            "locale": "de",
            "name": "ausdruck"
          },
          {
            "locale": "pt_BR",
            "name": "formula"
          }
        ],
        "description": "hidden direct roll of dice expression, e.g. `2d6`. Configuration with `/channel_config`",
        "descriptionLocales": [
          {
            "locale": "de",
            "description": "Versteckter direkter Wurf eines W체rfelausdruckes, z.B. `2d6`. Konfiguration mit `/channel_config`"
          },
          {
            "locale": "pt_BR",
            "description": "f처rmula de rolagem de dados direta escondida, e.g. `2d6`. Configure com `/channel_config`"
          }
        ],
        "required": true,
        "choices": [ ],
        "options": [ ],
        "autoComplete": false
      }
    ]
  }
]


de.janno.discord.bot.command.directRoll.HiddenRollCommandTest.getId=[
  "h"
]